src/edutap/wallet_google/clientpool.py

```diff
@@ -0,0 +1,234 @@
+        """Get or create a persistent async HTTP client from the pool.
+
+        Returns a cached AsyncAssertionClient for the given credentials. The same client
+        instance is reused across multiple API calls for optimal connection pooling.
+
+        The client is task-safe and can be shared across async tasks (within same event loop).
+
+        :param credentials: Client credentials as dict. If not given, credentials
+                            are read from file defined in settings.
+        :return:            The async assertion client (persistent).
+        """
+        credentials = self._get_credentials(credentials)
+        key = self._get_credentials_key(credentials)
+
+        # Thread-safe client creation (only create once per credentials)
+        with self._lock:
+            if key not in self._async_clients:
+                config = self._build_client_config(credentials)
+                # Note: AsyncAssertionClient doesn't support client_cls parameter for custom clients
+                self._async_clients[key] = AsyncAssertionClient(**config)
+
+        return self._async_clients[key]
+
+    def close_all_clients(self):
+        """Close all cached sync clients.
+
+        Call this method at application shutdown to properly close all
+        persistent clients and release resources.
+
+        This method is synchronous and closes only sync clients.
+        For async clients, use aclose_all_clients().
+        """
+        with self._lock:
+            for client in self._sync_clients.values():
+                client.close()
+            self._sync_clients.clear()
+
+    async def aclose_all_clients(self):
+        """Close all cached async clients.
+
+        Call this method at application shutdown to properly close all
+        persistent async clients and release resources.
+
+        This method must be called from an async context.
+        For sync clients, use close_all_clients().
+        """
+        # Don't need lock here since we're in async context
+        for client in list(self._async_clients.values()):
+            await client.aclose()
+        self._async_clients.clear()
+
+    def url(self, name: str, additional_path: str = "") -> str:
+        """
+        Create the URL for the CRUD operations.
+
+        :param name:            Registered name of the model.
+        :param additional_path: Append this to the path.
+                                Must start with a forward slash.
+
+        :return:                The url of the google RESTful API endpoint to handle
+                                this model
+        """
+        model_metadata = lookup_metadata_by_name(name)
+        return f"{self.settings.api_url}/{model_metadata['url_part']}{additional_path}"
+
+
+# Singleton instance for both sync and async operations
+client_pool = ClientPoolManager()
+
+
+# Register cleanup handler to close clients on process exit
```

Comment:
either that methode nees to be async and call both internal methods to close sync and async clients or the atexit register has to register both internal methods. 

The whole module level _cleanup_clients is not neccessary and atexit register can also move into the __init__ of CleintPoolManager and will that use the class scope

Reply:
documented cleanup for async (needs to be called in asyncloop, not in atexit.